### PR TITLE
Typing client file

### PR DESCRIPTION
**Pull request by enadeau**
## Description

Add type annotations to the client file. See #373

Still need to fix a couple small typing error and would be best to address #383
first as it relates to the test that are now failling.

## Type of change

Select the statement best describes this pull request.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update
- [x] Other (please describe)

## Does your PR include tests

If you are fixing a bug or adding a feature, we appreciate (but do not require)
tests to support whatever fix of feature you're implementing.

- [ ] Yes
- [ ] No, i'd like some help with tests
- [x] This change doesn't require tests

## Did you include your contribution to the change log?

- [x] Yes, `changelog.md` is up-to-date.

*If you are having a hard time getting the tests to run correctly feel free to
ping one of the maintainers here!*

### CODE REVIEW

**Comment by enadeau** [comment]
Back from summer holiday! Just finished typing the client file. Sorry for the big diff but at least after that we will have most of the outward facing method typed I think.

File: stravalib/client.py [hunk 1]
```diff
@@ -263,12 +294,8 @@ def get_activities(self, before=None, after=None, limit=None):
...
 
-        if before:
-            before = self._utc_datetime_to_epoch(before)
-
-        if after:
-            after = self._utc_datetime_to_epoch(after)
-
+        before_epoch = self._utc_datetime_to_epoch(before) if before else None
+        after_epoch = self._utc_datetime_to_epoch(after) if after else None
         params = dict(before=before, after=after)
```
Comment (jsamoocha) [inline stravalib/client.py:299]: I don't see where `before_epoch` or `after_epoc` are used. Can it be that these were meant to be set as part of the `params` dict?

File: stravalib/client.py [hunk 2]
```diff
@@ -1305,15 +1370,20 @@ def get_segment_efforts(
         )
 
     def explore_segments(
-        self, bounds, activity_type=None, min_cat=None, max_cat=None
-    ):
+        self,
+        bounds: tuple[float, float, float, float]
```
Comment (jsamoocha) [inline stravalib/client.py:1374]: I'm curious why this must be a tuple (instead of any indexable sequence (which I wouldn't know how to type, BTW)).

File: stravalib/client.py [hunk 3]
```diff
@@ -1909,23 +2022,23 @@ def _fill_buffer(self):
...
         return self.next()
 
-    def next(self):
-        """ """
+    def next(self) -> T:
         if self.limit and self._counter >= self.limit:
             self._eof()
         if not self._buffer:
             self._fill_buffer()
+        assert isinstance(self._buffer, collections.deque)
```
Comment (jsamoocha) [inline stravalib/client.py:2041]: Again, out of curiosity: why did you add this assertion?

File: stravalib/client.py [hunk 4]
```diff
@@ -1815,7 +1920,15 @@ def delete_subscription(self, subscription_id, client_id, client_secret):
...
-class BatchedResultsIterator(object):
+T = TypeVar("T", bound=BaseModel)
+
+
+class ResultFetcher(Protocol):
+    def __call__(self, *, page: int, per_page: int) -> Iterable[Any]:
+        ...
+
+
+class BatchedResultsIterator(Generic[T]):
```
Comment (jsamoocha) [inline stravalib/client.py:1931]: It's great that you turned this into a generic class 👍

**Review by jsamoocha — COMMENTED** [review]
Thanks a lot for this PR! Before merging, I'd like to be sure we're not changing any behavior regarding the `before` and `after` parameters I mentioned in a specific comment.

File: stravalib/client.py (hunk 1 above)
Comment (enadeau) [reply stravalib/client.py:299]: Could catch they should be passed as param. I'll fix that

File: stravalib/client.py (hunk 2 above)
Comment (enadeau) [reply stravalib/client.py:1374]: The tuple allows to incorporate the fact that there is exactly for float in the sequence, so that's why I chose it. Note that the function with still work if call with the list but the mypy will complain.

For a general indexable sequence you can use `typing.Sequence[float]`

File: stravalib/client.py (hunk 4 above)
Comment (enadeau) [reply stravalib/client.py:1931]: That really helps get good typing of the method that uses it 🙂

File: stravalib/client.py (hunk 3 above)
Comment (enadeau) [reply stravalib/client.py]: It tells mypy that here buffer cannot be None. I've actually changed to assert it is not None to make it clearer.

**Comment by enadeau** [comment]
I think it is now ready to go :100:

**Review by jsamoocha — APPROVED** [review]
Awesome, thanks for the explanations and fixes!
